### PR TITLE
[expo-updates][android] add method to read stringified requestHeaders

### PR DESCRIPTION
**Pull request by jkhales**
# Why

This allows us to set request headers such a `'expo-channel-name'` in the manifest by means of a stringified JSON in the metadata:

`"{\"expo-channel-name\":\"main\"}"`

Unlike the Plist in iOS, there is no dict like metadata object.

Talking with @esamelson , we  considered the structured field header formate, but went with stringified JSON as being easier for most developers to understand.

# Note

We give specific metadata to other headers like `runtime-version` since "we need the runtime version in other pieces of the client side besides just for setting the headers. we need to be able to decide which of the stored updates we want to run without talking to the server"

# How

Read string, convert to JSON, convert to `Map<String,String>`.

# Test Plan

Confirmed that header was sent on manifest requests.

### CODE REVIEW

**Review by esamelson — COMMENTED** [review]
this looks good -- my only comment is a conceptual one. since the fields in this particular json object all represent headers, do we even want to allow types other than strings? since this is a native configuration value, i wonder if it might be better to actually just crash right away if there are non-string values, rather than attempting to coerce them to strings (which might lead to unexpected/confusing behavior later on).

**Comment by jkhales** [comment]
> this looks good -- my only comment is a conceptual one. since the fields in this particular json object all represent headers, do we even want to allow types other than strings? since this is a native configuration value, i wonder if it might be better to actually just crash right away if there are non-string values, rather than attempting to coerce them to strings (which might lead to unexpected/confusing behavior later on).

I was trying to mimic the iOS behavior which is more permissive, but hearing you say it convinces me that the crash is better. Will rework.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.java [hunk 1]
```diff
@@ -40,6 +44,24 @@
...
+    JSONObject jsonObject = new JSONObject(stringifiedJSON);
+    Iterator<String> keys = jsonObject.keys();
+    Map<String, String> newMap = new HashMap<>();
+    while(keys.hasNext()){
+      String key = keys.next();
+      String val;
+      try{
+        val = (String) jsonObject.get(key);
+      } catch (ClassCastException e){
+        throw new Exception("The JSON must be string valued.");
```
Comment (ide) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.java:57]:
```suggestion
        throw new Exception("The values in the JSON object must be strings");
```

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/UpdatesUtilsInstrumentationTest.java [hunk 2]
```diff
@@ -27,4 +31,34 @@ public void testParseDateString_writtenTimezoneWithColon() throws ParseException
...
+    Map<String,String> emptyMap = getMapFromStringifiedJSON("{}");
+    Assert.assertEquals(emptyMap, new HashMap<>());
+  }
+
+  @Test
+  public void testGetMapFromStringifiedJSON_expectedFormat() throws Exception {
+    Map<String, String> expected = new HashMap<>();
+    expected.put("expo-channel-name","main");
+
+    Map<String,String> emptyMap = getMapFromStringifiedJSON("{'expo-channel-name':'main'}");
```
Comment (ide) [inline packages/expo-updates/android/src/androidTest/java/expo/modules/updates/UpdatesUtilsInstrumentationTest.java:46]: JSON strings need to be double-quoted... I'm surprised this works at all.
```suggestion
    Map<String,String> emptyMap = getMapFromStringifiedJSON("{\"expo-channel-name\":\"main\"}");
```

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/UpdatesUtilsInstrumentationTest.java [hunk 3]
```diff
@@ -27,4 +31,34 @@ public void testParseDateString_writtenTimezoneWithColon() throws ParseException
...
+  }
+
+  @Test(expected = Exception.class)
+  public void testGetMapFromStringifiedJSON_throwsNonStringValue() throws Exception {
+    getMapFromStringifiedJSON("{'expo-channel-name':['main']}");
+  }
+
+  @Test(expected = Exception.class)
+  public void testGetMapFromStringifiedJSON_throwsNonStringKey() throws Exception {
+    getMapFromStringifiedJSON("{7:['main']}");
```
Comment (ide) [inline packages/expo-updates/android/src/androidTest/java/expo/modules/updates/UpdatesUtilsInstrumentationTest.java:62]: Need to update all these tests too..

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.java [hunk 4]
```diff
@@ -40,6 +44,24 @@
...
   private static final String UPDATES_EVENT_NAME = "Expo.nativeUpdatesEvent";
 
+  public static Map<String,String> getMapFromStringifiedJSON(String stringifiedJSON) throws Exception {
+    JSONObject jsonObject = new JSONObject(stringifiedJSON);
+    Iterator<String> keys = jsonObject.keys();
+    Map<String, String> newMap = new HashMap<>();
+    while(keys.hasNext()){
+      String key = keys.next();
+      String val;
+      try{
```
Comment (esamelson) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.java:54]:
```suggestion
      try {
```

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.java [hunk 5]
```diff
@@ -40,6 +44,24 @@
...
+        val = (String) jsonObject.get(key);
+      } catch (ClassCastException e){
```
Comment (esamelson) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.java:56]:
```suggestion
      } catch (ClassCastException e) {
```

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.java [hunk 6]
```diff
@@ -40,6 +44,24 @@
   private static final String UPDATES_DIRECTORY_NAME = ".expo-internal";
   private static final String UPDATES_EVENT_NAME = "Expo.nativeUpdatesEvent";
 
+  public static Map<String,String> getMapFromStringifiedJSON(String stringifiedJSON) throws Exception {
+    JSONObject jsonObject = new JSONObject(stringifiedJSON);
+    Iterator<String> keys = jsonObject.keys();
+    Map<String, String> newMap = new HashMap<>();
+    while(keys.hasNext()){
```
Comment (esamelson) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.java:51]:
```suggestion
    while (keys.hasNext()) {
```

**Review by esamelson — APPROVED** [review]
> I was trying to mimic the iOS behavior which is more permissive, but hearing you say it convinces me that the crash is better. Will rework.

That's a good point; let's plan to rework the iOS module to match this behavior.

My only other thought is perhaps we should name this method something that communicates more clearly that it has a specific purpose, or that it will throw if it encounters non-string values in the JSON object. Maybe something like `getHeadersMapFromJSONString`?

But lgtm otherwise!

**Pull request merged**